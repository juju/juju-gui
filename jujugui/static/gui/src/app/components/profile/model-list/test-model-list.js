/* Copyright (C) 2017 Canonical Ltd. */
'use strict';

const React = require('react');
const enzyme = require('enzyme');

const ProfileModelList = require('./model-list');
const Spinner = require('../../spinner/spinner');

describe('Profile Model List', function() {
  // JSON dump of a test listModelsWithInfo call.
  // The first two records are owned by 'tester'.
  // The last two are shared with 'tester'.
  const rawModelData = `[{
    "id": "2f929db7-08a1-4a75-8733-3a0352a6e9f5",
    "name": "mymodel",
    "series": "xenial",
    "provider": "ec2",
    "uuid": "2f929db7-08a1-4a75-8733-3a0352a6e9f5",
    "agentVersion": "",
    "sla": "",
    "slaOwner": "",
    "status": "available",
    "statusInfo": "",
    "controllerUUID": "a030379a-940f-4760-8fce-3062b41a04e9",
    "owner": "tester@external",
    "credential": "aws_tester@external_base",
    "credentialName": "base",
    "region": "eu-west-1",
    "cloud": "aws",
    "numMachines": 0,
    "users": [{
      "name": "tester@external",
      "displayName": "tester",
      "domain": "Ubuntu SSO",
      "lastConnection": "2017-07-06T14:47:03.000Z",
      "access": "admin"
    }, {
      "name": "tester2@external",
      "displayName": "tester2",
      "domain": "Ubuntu SSO",
      "lastConnection": "2017-05-25T18:53:35.000Z",
      "access": "read"
    }],
    "life": "alive",
    "isAlive": true,
    "isController": false,
    "lastConnection": "2017-07-06T14:47:03.000Z"
  }, {
    "id": "f507b811-d706-427e-8f01-d485fgc90a67",
    "name": "mymodel2",
    "series": "xenial",
    "provider": "gce",
    "uuid": "f507b811-d706-427e-8f01-d485fgc90a67",
    "agentVersion": "",
    "sla": "",
    "slaOwner": "",
    "status": "available",
    "statusInfo": "",
    "controllerUUID": "a030379a-940f-4760-8fce-3062b41a04e9",
    "owner": "tester@external",
    "credential": "google_tester@external_admin",
    "credentialName": "admin",
    "region": "us-central1",
    "cloud": "google",
    "numMachines": 0,
    "users": [{
      "name": "tester@external",
      "displayName": "tester",
      "domain": "Ubuntu SSO",
      "lastConnection": "2017-07-06T16:26:47.000Z",
      "access": "admin"
    }],
    "life": "alive",
    "isAlive": true,
    "isController": false,
    "lastConnection": "2017-07-06T16:26:47.000Z"
  }, {
    "id": "06b11eb5-abd6-48f3-8910-b54cf5945e60",
    "name": "test-db",
    "series": "xenial",
    "provider": "gce",
    "uuid": "06b11eb5-abd6-48f3-8910-b54cf5945e60",
    "agentVersion": "",
    "sla": "",
    "slaOwner": "",
    "status": "available",
    "statusInfo": "",
    "controllerUUID": "a030379a-940f-4760-8fce-3062b41a04e9",
    "owner": "tester2@external",
    "credential": "google_tester2@external_google-dbtest",
    "credentialName": "google-dbtest",
    "region": "us-east1",
    "cloud": "google",
    "numMachines": 2,
    "users": [{
      "name": "tester@external",
      "displayName": "tester",
      "domain": "Ubuntu SSO",
      "lastConnection": null,
      "access": "read"
    }, {
      "name": "tester2@external",
      "displayName": "tester2",
      "domain": "Ubuntu SSO",
      "lastConnection": null,
      "access": "admin"
    }],
    "life": "alive",
    "isAlive": true,
    "isController": false,
    "lastConnection": null
  }, {
    "id": "509f6e4c-4da4-49c8-8f19-537c33b4d3a0",
    "name": "website",
    "series": "xenial",
    "provider": "gce",
    "uuid": "509f6e4c-4da4-49c8-8f19-537c33b4d3a0",
    "agentVersion": "",
    "sla": "",
    "slaOwner": "",
    "status": "available",
    "statusInfo": "",
    "controllerUUID": "a030379a-940f-4760-8fce-3062b41a04e9",
    "owner": "tester3@external",
    "credential": "google_tester3@external_gce",
    "credentialName": "gce",
    "region": "us-east1",
    "cloud": "google",
    "numMachines": 25,
    "users": [{
      "name": "tester4@external",
      "displayName": "tester4",
      "domain": "Ubuntu SSO",
      "lastConnection": "2017-06-09T16:24:28.000Z",
      "access": "read"
    }, {
      "name": "tester@external",
      "displayName": "tester",
      "domain": "Ubuntu SSO",
      "lastConnection": "2017-07-05T01:42:05.000Z",
      "access": "admin"
    }],
    "life": "alive",
    "isAlive": true,
    "isController": false,
    "lastConnection": "2017-07-05T01:42:05.000Z"
  }]`;

  const renderComponent = (options = {}) =>
    enzyme.shallow(
      <ProfileModelList
        acl={{}}
        addNotification={options.addNotification || sinon.stub()}
        baseURL="/gui/"
        changeState={options.changeState || sinon.stub()}
        destroyModel={options.destroyModel || sinon.stub()}
        facadesExist={true}
        listModelsWithInfo={
          options.listModelsWithInfo ||
          sinon.stub().callsArgWith(0, null, JSON.parse(rawModelData))
        }
        switchModel={options.switchModel || sinon.stub()}
        userInfo={options.userInfo || { profile: 'tester' }} />
    );

  it('can render', () => {
    const wrapper = renderComponent({ gisf: true });
    expect(wrapper).toMatchSnapshot();
  });

  it('does not break for superusers', () => {
    // Users with access to all models but no models of their own, or shared with them.
    const models = JSON.parse(rawModelData);
    models.push(
      JSON.parse(`{
      "id": "2f929db7-08a1-4a75-8733-3a0352a6e9f5",
      "name": "mymodel-foo",
      "series": "xenial",
      "provider": "ec2",
      "uuid": "2f929db7-08a1-4a75-8733-3a0352a6e9f5",
      "agentVersion": "",
      "sla": "",
      "slaOwner": "",
      "status": "available",
      "statusInfo": "",
      "controllerUUID": "a030379a-940f-4760-8fce-3062b41a04e9",
      "owner": "'somesuperuser'@external",
      "credential": "aws_tester@external_base",
      "credentialName": "base",
      "region": "eu-west-1",
      "cloud": "aws",
      "numMachines": 0,
      "users": [{
        "name": "somesuperuser@external",
        "displayName": "somesuperuser",
        "domain": "Ubuntu SSO",
        "lastConnection": "2017-07-06T14:47:03.000Z",
        "access": "admin"
      }],
      "life": "alive",
      "isAlive": true,
      "isController": false,
      "lastConnection": "2017-07-06T14:47:03.000Z"
    }`)
    );
    const wrapper = renderComponent({
      listModelsWithInfo: sinon.stub().callsArgWith(0, null, models),
      userInfo: { profile: 'somesuperuser' }
    });
    // It should only show the single model that they explicitly own.
    assert.equal(
      wrapper
        .find('.profile__title-count')
        .html()
        .includes('(1)'),
      true
    );
  });

  it('can render without any models', () => {
    const wrapper = renderComponent({
      listModelsWithInfo: sinon.stub().callsArgWith(0, null, null)
    });
    assert.equal(
      wrapper
        .find('.profile__title-count')
        .html()
        .includes('(0)'),
      true
    );
    assert.equal(wrapper.find('BasicTable').length, 0);
  });

  it('does not break with model data in an unexpected format', () => {
    const wrapper = renderComponent({
      listModelsWithInfo: sinon.stub().callsArgWith(0, null, [''])
    });
    assert.equal(
      wrapper
        .find('.profile__title-count')
        .html()
        .includes('(0)'),
      true
    );
    assert.equal(wrapper.find('BasicTable').length, 0);
  });

  it('does not show the trash icon for controller models', () => {
    const models = JSON.parse(rawModelData).slice(0, 1);
    models[0].isController = true;
    const wrapper = renderComponent({
      listModelsWithInfo: sinon.stub().callsArgWith(0, null, models)
    });
    assert.strictEqual(wrapper.find('BasicTable').prop('rows')[0].columns[5].content, null);
  });

  it('does not show models that are being destroyed', () => {
    const models = [{ isAlive: false }];
    models.push(
      JSON.parse(`{
      "id": "2f929db7-08a1-4a75-8733-3a0352a6e9f5",
      "name": "mymodel",
      "series": "xenial",
      "provider": "ec2",
      "uuid": "2f929db7-08a1-4a75-8733-3a0352a6e9f5",
      "agentVersion": "",
      "sla": "",
      "slaOwner": "",
      "status": "available",
      "statusInfo": "",
      "controllerUUID": "a030379a-940f-4760-8fce-3062b41a04e9",
      "owner": "tester@external",
      "credential": "aws_tester@external_base",
      "credentialName": "base",
      "region": "eu-west-1",
      "cloud": "aws",
      "numMachines": 0,
      "users": [{
        "name": "tester@external",
        "displayName": "tester",
        "domain": "Ubuntu SSO",
        "lastConnection": "2017-07-06T14:47:03.000Z",
        "access": "admin"
      }, {
        "name": "tester2@external",
        "displayName": "tester2",
        "domain": "Ubuntu SSO",
        "lastConnection": "2017-05-25T18:53:35.000Z",
        "access": "read"
      }],
      "life": "alive",
      "isAlive": true,
      "isController": false,
      "lastConnection": "2017-07-06T14:47:03.000Z"
    }`)
    );
    const wrapper = renderComponent({
      listModelsWithInfo: sinon.stub().callsArgWith(0, null, models)
    });
    assert.equal(
      wrapper
        .find('.profile__title-count')
        .html()
        .includes('(1)'),
      true
    );
  });

  it('displays an error when destroying a model fails', () => {
    const addNotification = sinon.stub();
    const wrapper = renderComponent({
      addNotification: addNotification,
      destroyModel: (modelUUID, callback) => {
        callback(['Error 1', 'Error 2'], {});
      }
    });
    wrapper.instance()._confirmDestroy('test');
    assert.equal(addNotification.callCount, 2);
  });

  it('displays a spinner when loading', () => {
    const wrapper = renderComponent({
      listModelsWithInfo: sinon.stub()
    });
    wrapper.update();
    const expected = (
      <div className="profile-model-list">
        <Spinner />
      </div>
    );
    assert.compareJSX(wrapper, expected);
  });

  it('switches to a model that has been clicked on', () => {
    const switchModel = sinon.stub();
    const changeState = sinon.stub();
    const e = {
      preventDefault: sinon.stub(),
      stopPropagation: sinon.stub()
    };
    const wrapper = renderComponent({
      switchModel: switchModel,
      changeState: changeState
    });
    wrapper.update();
    const link = wrapper.find('BasicTable').prop('rows')[0].columns[0].content;
    link.props.onClick(e);
    assert.equal(e.preventDefault.callCount, 1);
    assert.equal(e.stopPropagation.callCount, 1);
    assert.equal(changeState.callCount, 1, 'changeState not called');
    assert.deepEqual(changeState.args[0], [{ profile: null }]);
    assert.equal(switchModel.callCount, 1, 'switchModel not called');
    assert.deepEqual(switchModel.args[0], [
      {
        name: 'mymodel',
        id: '2f929db7-08a1-4a75-8733-3a0352a6e9f5',
        owner: 'tester'
      }
    ]);
  });
});
