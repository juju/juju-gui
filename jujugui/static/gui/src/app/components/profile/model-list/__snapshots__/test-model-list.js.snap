// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile Model List can render 1`] = `
<div
  className="profile-model-list"
>
  <div
    className="profile-model-list__header v1"
  >
    <CreateModelButton
      changeState={[Function]}
      switchModel={[Function]}
      title="Start a new model"
      type="inline-neutral"
    />
    <h2
      className="profile__title"
    >
      My models
      <span
        className="profile__title-count"
      >
        (
        4
        )
      </span>
    </h2>
  </div>
  <BasicTable
    headerClasses={
      Array [
        "profile__entity-table-header-row",
      ]
    }
    headerColumnClasses={
      Array [
        "profile__entity-table-header-column",
      ]
    }
    headers={
      Array [
        Object {
          "columnSize": 3,
          "content": "Name",
        },
        Object {
          "columnSize": 2,
          "content": "Owner",
        },
        Object {
          "columnSize": 3,
          "content": "Machines, cloud/region",
        },
        Object {
          "columnSize": 1,
          "content": "",
        },
        Object {
          "columnSize": 2,
          "content": "Last accessed",
        },
        Object {
          "columnSize": 1,
          "content": "",
        },
      ]
    }
    rowClasses={
      Array [
        "profile__entity-table-row",
      ]
    }
    rowColumnClasses={
      Array [
        "profile__entity-table-column",
      ]
    }
    rows={
      Array [
        Object {
          "columns": Array [
            Object {
              "columnSize": 3,
              "content": <a
                href="/gui/u/tester/mymodel"
                onClick={[Function]}
              >
                mymodel
              </a>,
            },
            Object {
              "columnSize": 2,
              "content": "Me",
            },
            Object {
              "columnSize": 3,
              "content": <div>
                <span
                  className="profile-model-list__machine-number"
                >
                  0
                </span>
                aws
                /eu-west-1
              </div>,
            },
            Object {
              "columnSize": 1,
              "content": <div
                className="profile-model-list__access tooltip"
              >
                <span
                  className="tooltip__tooltip"
                >
                  <span
                    className="tooltip__inner tooltip__inner--down"
                  >
                    admin
                  </span>
                </span>
                <SvgIcon
                  name="user_16"
                  size="16"
                />
              </div>,
            },
            Object {
              "columnSize": 2,
              "content": <DateDisplay
                date="2017-07-06T14:47:03.000Z"
                relative={true}
              />,
            },
            Object {
              "classes": Array [
                "u-text-align--right",
              ],
              "columnSize": 1,
              "content": <a
                onClick={[Function]}
              >
                <SvgIcon
                  name="delete_16"
                  size="16"
                />
              </a>,
            },
          ],
          "expandedContent": <div
            className="profile-model-list__expanded-content"
          >
            <div
              className="three-col"
            >
              <a
                href="/gui/u/tester/mymodel"
                onClick={[Function]}
              >
                mymodel
              </a>
            </div>
            <div
              className="two-col"
            >
              Me
            </div>
            <div
              className="three-col"
            >
              <div>
                <span
                  className="profile-model-list__machine-number"
                >
                  0
                </span>
                aws
                /eu-west-1
              </div>
            </div>
            <div
              className="one-col"
            >
              <div
                className="profile-model-list__access tooltip"
              >
                <span
                  className="tooltip__tooltip"
                >
                  <span
                    className="tooltip__inner tooltip__inner--down"
                  >
                    admin
                  </span>
                </span>
                <SvgIcon
                  name="user_16"
                  size="16"
                />
              </div>
            </div>
            <div
              className="two-col"
            >
              <DateDisplay
                date="2017-07-06T14:47:03.000Z"
                relative={true}
              />
            </div>
            <div
              className="one-col last-col u-text-align--right"
            >
              <a
                onClick={[Function]}
              >
                <SvgIcon
                  name="delete_16"
                  size="16"
                />
              </a>
            </div>
            <div
              className="three-col prepend-five profile-model-list__credential-name"
            >
              <span
                className="link"
                onClick={[Function]}
                role="button"
                tabIndex="0"
              >
                base
              </span>
            </div>
          </div>,
          "key": "mymodel",
        },
        Object {
          "columns": Array [
            Object {
              "columnSize": 3,
              "content": <a
                href="/gui/u/tester/mymodel2"
                onClick={[Function]}
              >
                mymodel2
              </a>,
            },
            Object {
              "columnSize": 2,
              "content": "Me",
            },
            Object {
              "columnSize": 3,
              "content": <div>
                <span
                  className="profile-model-list__machine-number"
                >
                  0
                </span>
                google
                /us-central1
              </div>,
            },
            Object {
              "columnSize": 1,
              "content": <div
                className="profile-model-list__access tooltip"
              >
                <span
                  className="tooltip__tooltip"
                >
                  <span
                    className="tooltip__inner tooltip__inner--down"
                  >
                    admin
                  </span>
                </span>
                <SvgIcon
                  name="user_16"
                  size="16"
                />
              </div>,
            },
            Object {
              "columnSize": 2,
              "content": <DateDisplay
                date="2017-07-06T16:26:47.000Z"
                relative={true}
              />,
            },
            Object {
              "classes": Array [
                "u-text-align--right",
              ],
              "columnSize": 1,
              "content": <a
                onClick={[Function]}
              >
                <SvgIcon
                  name="delete_16"
                  size="16"
                />
              </a>,
            },
          ],
          "expandedContent": <div
            className="profile-model-list__expanded-content"
          >
            <div
              className="three-col"
            >
              <a
                href="/gui/u/tester/mymodel2"
                onClick={[Function]}
              >
                mymodel2
              </a>
            </div>
            <div
              className="two-col"
            >
              Me
            </div>
            <div
              className="three-col"
            >
              <div>
                <span
                  className="profile-model-list__machine-number"
                >
                  0
                </span>
                google
                /us-central1
              </div>
            </div>
            <div
              className="one-col"
            >
              <div
                className="profile-model-list__access tooltip"
              >
                <span
                  className="tooltip__tooltip"
                >
                  <span
                    className="tooltip__inner tooltip__inner--down"
                  >
                    admin
                  </span>
                </span>
                <SvgIcon
                  name="user_16"
                  size="16"
                />
              </div>
            </div>
            <div
              className="two-col"
            >
              <DateDisplay
                date="2017-07-06T16:26:47.000Z"
                relative={true}
              />
            </div>
            <div
              className="one-col last-col u-text-align--right"
            >
              <a
                onClick={[Function]}
              >
                <SvgIcon
                  name="delete_16"
                  size="16"
                />
              </a>
            </div>
            <div
              className="three-col prepend-five profile-model-list__credential-name"
            >
              <span
                className="link"
                onClick={[Function]}
                role="button"
                tabIndex="0"
              >
                admin
              </span>
            </div>
          </div>,
          "key": "mymodel2",
        },
        Object {
          "columns": Array [
            Object {
              "columnSize": 3,
              "content": <a
                href="/gui/u/tester2/test-db"
                onClick={[Function]}
              >
                test-db
              </a>,
            },
            Object {
              "columnSize": 2,
              "content": "tester2",
            },
            Object {
              "columnSize": 3,
              "content": <div>
                <span
                  className="profile-model-list__machine-number"
                >
                  2
                </span>
                google
                /us-east1
              </div>,
            },
            Object {
              "columnSize": 1,
              "content": <div
                className="profile-model-list__access tooltip"
              >
                <span
                  className="tooltip__tooltip"
                >
                  <span
                    className="tooltip__inner tooltip__inner--down"
                  >
                    read
                  </span>
                </span>
                <SvgIcon
                  name="show_16"
                  size="16"
                />
              </div>,
            },
            Object {
              "columnSize": 2,
              "content": <DateDisplay
                date="--"
                relative={true}
              />,
            },
            Object {
              "classes": Array [
                "u-text-align--right",
              ],
              "columnSize": 1,
              "content": null,
            },
          ],
          "expandedContent": <div
            className="profile-model-list__expanded-content"
          >
            <div
              className="three-col"
            >
              <a
                href="/gui/u/tester2/test-db"
                onClick={[Function]}
              >
                test-db
              </a>
            </div>
            <div
              className="two-col"
            >
              tester2
            </div>
            <div
              className="three-col"
            >
              <div>
                <span
                  className="profile-model-list__machine-number"
                >
                  2
                </span>
                google
                /us-east1
              </div>
            </div>
            <div
              className="one-col"
            >
              <div
                className="profile-model-list__access tooltip"
              >
                <span
                  className="tooltip__tooltip"
                >
                  <span
                    className="tooltip__inner tooltip__inner--down"
                  >
                    read
                  </span>
                </span>
                <SvgIcon
                  name="show_16"
                  size="16"
                />
              </div>
            </div>
            <div
              className="two-col"
            >
              <DateDisplay
                date="--"
                relative={true}
              />
            </div>
            <div
              className="one-col last-col u-text-align--right"
            />
            <div
              className="twelve-col"
            >
              No additional information available on shared model.
            </div>
          </div>,
          "key": "test-db",
        },
        Object {
          "columns": Array [
            Object {
              "columnSize": 3,
              "content": <a
                href="/gui/u/tester3/website"
                onClick={[Function]}
              >
                website
              </a>,
            },
            Object {
              "columnSize": 2,
              "content": "tester3",
            },
            Object {
              "columnSize": 3,
              "content": <div>
                <span
                  className="profile-model-list__machine-number"
                >
                  25
                </span>
                google
                /us-east1
              </div>,
            },
            Object {
              "columnSize": 1,
              "content": <div
                className="profile-model-list__access tooltip"
              >
                <span
                  className="tooltip__tooltip"
                >
                  <span
                    className="tooltip__inner tooltip__inner--down"
                  >
                    admin
                  </span>
                </span>
                <SvgIcon
                  name="user_16"
                  size="16"
                />
              </div>,
            },
            Object {
              "columnSize": 2,
              "content": <DateDisplay
                date="2017-07-05T01:42:05.000Z"
                relative={true}
              />,
            },
            Object {
              "classes": Array [
                "u-text-align--right",
              ],
              "columnSize": 1,
              "content": <a
                onClick={[Function]}
              >
                <SvgIcon
                  name="delete_16"
                  size="16"
                />
              </a>,
            },
          ],
          "expandedContent": <div
            className="profile-model-list__expanded-content"
          >
            <div
              className="three-col"
            >
              <a
                href="/gui/u/tester3/website"
                onClick={[Function]}
              >
                website
              </a>
            </div>
            <div
              className="two-col"
            >
              tester3
            </div>
            <div
              className="three-col"
            >
              <div>
                <span
                  className="profile-model-list__machine-number"
                >
                  25
                </span>
                google
                /us-east1
              </div>
            </div>
            <div
              className="one-col"
            >
              <div
                className="profile-model-list__access tooltip"
              >
                <span
                  className="tooltip__tooltip"
                >
                  <span
                    className="tooltip__inner tooltip__inner--down"
                  >
                    admin
                  </span>
                </span>
                <SvgIcon
                  name="user_16"
                  size="16"
                />
              </div>
            </div>
            <div
              className="two-col"
            >
              <DateDisplay
                date="2017-07-05T01:42:05.000Z"
                relative={true}
              />
            </div>
            <div
              className="one-col last-col u-text-align--right"
            >
              <a
                onClick={[Function]}
              >
                <SvgIcon
                  name="delete_16"
                  size="16"
                />
              </a>
            </div>
            <div
              className="twelve-col"
            >
              No additional information available on shared model.
            </div>
          </div>,
          "key": "website",
        },
      ]
    }
  />
</div>
`;
